HP-318: step 5: basic data editing

NikoHelle opened the pull request:
Added a HTML form to BasicData-component and buttons to enter edit mode and save changes. New mutation hook for saving data. Data update functions added to editData. 

A lot of new tests and mock updates.

NikoHelle committed:
updateMutationVariables to support partial updates

NikoHelle committed:
updateMutationVariables to support partial updates

NikoHelle committed:
Update jest-fetch-mock
Now MockApolloClient can return query variables

NikoHelle committed:
Add AnyObject type to typings

NikoHelle committed:
Add all profile data to UpdateMyProfile query
This way cache can be fully updated without a new MyProfile query
Generated also generatedTypes.ts

NikoHelle committed:
Add all profile data to UpdateMyProfile query
This way cache can be fully updated without a new MyProfile query
Generated also generatedTypes.ts

NikoHelle committed:
Add hook for mutations

The main purpose of this commit is "useProfileMutations.ts". It sends profile data updates to the server and updates cache after that. It uses ProfileContext and also passes through its data, so "useProfileDataEditor" is required to use only useProfileMutations and not also useProfileContext.

Testing mutations require a lot of new testing tools:
- "myProfileMocking.ts" has "cloneAndManipulateProfile" that provides functions for easy profile editing. It outputs new profile and also formValues needed to mock updates. New profile data is then used in mock responses and formValues when creating update variables.
- "componentMocking.ts" exposes the "useProfileMutations" hook for testing. Added also simple "createResultPropertyTracker" that creates a promise based listener for hook's properties.  "ProfileContextFetcher" component just auto-fetches profile data for the hook.

Also bundled all clean-up functions for tests to "cleanComponentMocks" function and updated ProfileContext to use it like the new UseProfileMutations.tests.ts

useProfileDataEditor hook now uses useProfileMutations instead of useProfileContext

NikoHelle committed:
Add await-to-js
Makes async/await error handling easier

NikoHelle committed:
Added data updating to editData.ts

Renamed 'create' to 'createNewItem' to make it cleared what it creates

New functions:
- updateItemAndCreateSaveData: called when saving. It updates passed item by cloning it and all other items too. Value of the cloned item is set as passed new value. Function returns a formValue object that is passed on to updateMutationVariables which creates the update variables. While saving the item may not change and "preventDoubleEdits" -function checks that.

- updateData: called after save. If editData is waiting for an update (any item.saving is set). Items are updated if new profile data has same value edit data does.

-updateAfterSavingError: when save error occurs, item.saving is set to "undefined"

-updateItemAndCloneList: clones item, updates clone.value and returns all items as new, cloned list

-hasItemUpdated: returns true/false, depending on whether prop that is being saved (value/primary) matches the same prop in profile

-updateItems: updates items being saved if hasItemUpdated() returns true. Tracks also the new item (even they cannot be created yet). There can be only one new item at a time, so once it's found in profile data, no need to check new items anymore. Returns new list only if any items where updated.

NikoHelle committed:
Added saving to useProfileDataEditor

useProfileDataEditor now uses new update functions in editData and useProfileMutations to update profile

NikoHelle committed:
Editing and saving added to BasicData.tsx

Basic data can now be edited.

New components:

- SaveIndicator.tsx. Notifies screen readers that save is in progress
- EditingNotifications.tsx Notifies all users when saving in complete (success/error)
- useNotificationContent.ts creates content for EditingNotifications.tsx and makes it more general and easier to update.
- EditButton.tsx renders buttons for making changes (edit/remove/set-primary)
- FormButtons.tsx renders save/cancel buttons when editing

Updates
- componentMocking has a lot of new helpers to test components and UI-flow.
- BasicData.test.tsx updated of course

NikoHelle committed:
Cancel-button resets to previous version

When user edits an item, tries to save it and it fails, item.value should be reset to previous value. Added storage for previous versions: backups. Before storing new value, old one is backed up and restored with reset().

NikoHelle committed:
Changed Profile API dev address
.env.development updated.
env-config.js is an auto-generated file.

NikoHelle committed:
Fix wrong button position.

VilleMiekkoja commented on src/common/test/MockApolloClientProvider.tsx:
```diff
@@ -23,8 +25,13 @@ export function MockApolloClientProvider({
   const customGlobal = global as GlobalWithFetchMock;
-  customGlobal.fetchMock.mockResponse((variables?: RequestVariables) => {
-    const response = createMockedProfileResponse(responseProvider(variables));
+  customGlobal.fetchMock.mockResponse(async (req: Request) => {
```
Does this mock all fetch calls? Is there a risk that this mock stays on for some tests that would not wish to mock fetch?

NikoHelle replied:
Yes, it mocks all calls - but only when the test includes the MockApolloClientProvider.  And only Apollo graphQL-queries are fetched in the UI. It is true that other fetches are mocked too, but at the moment there aren't any and then MockApollo should not be used. Is the risk too big?

VilleMiekkoja replied:
Probably not a big risk, just wondering if it could cause problems in the future where developer would not notice why is his fetch mocked if he tries testing code with other fetches.

NikoHelle replied:
I'll change this to catch only /graphql/ paths.

NikoHelle replied:
Done.

VilleMiekkoja commented on src/common/test/componentMocking.tsx:
```diff
@@ -81,8 +127,8 @@ export const renderProfileContextWrapper = async (
-  const waitForDataChange: TestTools['waitForDataChange'] = async () => {
-    const latest = getLastTime('dataUpdateTime');
+  const waitForDataChange: TestTools['waitForDataChange'] = async previousChangeTime => {
+    const latest = previousChangeTime || getLastTime('dataUpdateTime');
```
Minor improvement suggestion to help readability: Could the getLastTime function be defined little closer where it is applied? Seems to be also applied inside submit function, but maybe we could put the definition above this waitForDataChange?

NikoHelle replied:
Referring to another re-ordering comment:
https://github.com/City-of-Helsinki/open-city-profile-ui/pull/154#discussion_r618158453

VilleMiekkoja commented on src/common/test/componentMocking.tsx:
```diff
@@ -17,26 +22,67 @@ import {
+  resetApolloMocks();
 };
 
 export const emptyResponseProvider: ResponseProvider = () => ({});
```
Off-topic comment for line 90: the renderProfileContextWrapper seems to do more than just rendering; it also provides various test utility functions. Can't figure for a better name though. Maybe it is doing too much?

NikoHelle replied:
Renamed to renderComponentWithMocksAndContexts. The return type in TS-files should be enough to tell what a function returns instead of using renderComponentWithMocksAndContextsAndReturnTestTools 😄 

All tested components need Router, mocked GraphQL, Profile Context, and Toast Context. And ability to fetch data, find elements, interact with DOM and waitFor something to be complete. This test helper returns context and tools for those. 

Plan B would be to render components in one func and pass result of that to another function(s) to do the rest. In each test file.

VilleMiekkoja commented on src/common/test/componentMocking.tsx:
```diff
@@ -97,14 +143,33 @@ export const renderProfileContextWrapper = async (
   };
 
-  const getElement: TestTools['getElement'] = ({ testId, text }) => {
+  const getElement: TestTools['getElement'] = ({
```
It seems that this function only wants one of the arguments per function call? Im not sure if it's easy to tell from the api of the function, that user should only pass one of them? One option could be to separate the qualifier and value. Something like

`const getElementByQualifier(qualifier, value)`. Though then we would need to type all the qualifiers, and that would also have an impact to the readability probably. 🤔

NikoHelle replied:
In normal use, why would you use multiple selectors at the same time? Like id + testid, label + text or id  + text. QuerySelector with label/text/value would make sense, but others not. When selector with value is needed there is waitForElementAndValue.
Selector with label/text is not currently needed.

I added this so there is only one function to get an element.  getElement is also used in tests outside of this file. If it did not exists, all funcs from testing-library should be passed through to tests
getByLabelText(label);
getByTestId(testId);
getByText(text);
getByDisplayValue(valueSelector);
(+ create new ones for id and querySelector )

GetElement is easier to use that than finding each time the correct one from testing-library.

VilleMiekkoja commented on src/common/test/componentMocking.tsx:
```diff
@@ -119,17 +184,74 @@ export const renderProfileContextWrapper = async (
+
+  const waitForElementAndValue: TestTools['waitForElementAndValue'] = async props => {
+    const { selector, value } = props;
+    const elementValue = await getTextOrInputValue(selector);
```
Minor improvement suggestion: The getTextOrInputValue function seems to be applied only inside this function, and also it is inside the return promise. I wonder if we could define it above this waitForElementAndValue function so it would be closer to usage?

NikoHelle replied:
Referring to another re-ordering comment:
https://github.com/City-of-Helsinki/open-city-profile-ui/pull/154#discussion_r618158453

VilleMiekkoja commented on src/common/test/componentMocking.tsx:
```diff
@@ -119,17 +184,74 @@ export const renderProfileContextWrapper = async (
     return Promise.resolve(value);
   };
 
+  const triggerAction: TestTools['triggerAction'] = async selector => {
```
Would clickElement describe this function better?

NikoHelle replied:
Done.

VilleMiekkoja commented on src/common/cssHelpers/form.module.css:
```diff
@@ -139,7 +139,7 @@
   
```
dev.hel.fi recommends _camelCase_ for style file names: https://dev.hel.fi/coding-standards#naming

NikoHelle replied:
All old profile css files are named *.module.css. Probably because name is required so React identifies them as css modules. 

https://create-react-app.dev/docs/adding-a-css-modules-stylesheet/

VilleMiekkoja commented on src/common/test/componentMocking.tsx:
```diff
@@ -232,6 +354,74 @@ export const exposeProfileContext = (
+  const [fetchStarted, setFetchStarted] = useState(false);
+  const { data, fetch } = useContext(ProfileContext);
+  if (!fetchStarted) {
+    fetch();
```
Data fetching side effects are usually wrapped inside useEffect hook. Is it not needed this time?

NikoHelle replied:
The idea of this component is to start fetching and only once. If useEffect was used, "fetch" must be hook's dependency and I can't guarantee if the Apollo's fetch function never changes between renders. Change would re-run the useEffect and fetch.

VilleMiekkoja commented on src/common/test/MockApolloClientProvider.tsx:
```diff
@@ -23,8 +25,13 @@ export function MockApolloClientProvider({
   responseProvider: ResponseProvider;
 }): React.ReactElement {
   const customGlobal = global as GlobalWithFetchMock;
```
I see that the jest-fetch-mock library is updated in this pr from 2.1.2 to 3.0.3 version. The documentation of 3.0.3 jest-fetch-mock (https://www.npmjs.com/package/jest-fetch-mock) no longer mention this type _GlobalWithFetchMock_ . So I wonder if the preferred way of handling globals with the jest-fetch-mock library has changed?

NikoHelle replied:
GlobalWithFetchMock still exists, but you're right, documentation has changed regarding the setup. I'll try to remove  GlobalWithFetchMock and use enableMocks()

NikoHelle replied:
Global stuff removed!

VilleMiekkoja commented on src/common/test/myProfileMocking.ts:
```diff
@@ -155,3 +207,189 @@ export const getVerifiedData = (
+  return dataType === 'emails'
+    ? 'EmailNodeEdge'
+    : dataType === 'phones'
+    ? 'PhoneNodeEdge'
```
Personally I find double ternary little confusing, but could be just that I haven't got to used to it. Would probably prefer regular if else for clarity

NikoHelle replied:
Done.

VilleMiekkoja commented on src/common/test/myProfileMocking.ts:
```diff
@@ -155,3 +207,189 @@ export const getVerifiedData = (
+  if (newPrimary) {
+    newPrimary.primary = true;
+  }
+  const propName = getPrimaryPropName(dataType);
```
The getPrimaryPropName function seems to be applied only inside this setPrimaryNode function. It might increase readability if it would be defined straight above this setPrimaryNode function

NikoHelle replied:
Referring to another re-ordering comment:
https://github.com/City-of-Helsinki/open-city-profile-ui/pull/154#discussion_r618158453

VilleMiekkoja commented on src/common/test/myProfileMocking.ts:
```diff
@@ -155,3 +207,189 @@ export const getVerifiedData = (
+      return currentProfile;
+    },
+    add(dataType, data, index = -1) {
+      const edge = createEdge(dataType, data);
```
The createEdge function seems to be applied only inside this cloneAndManipulateProfile function. Would it increase readability to have it defined straight above this cloneAndManipulateProfile function so it would be closer to usage? Though the setPrimaryNode function is also applied inside this cloneAndManipulateProfile function.

NikoHelle replied:
Referring to another re-ordering comment:
https://github.com/City-of-Helsinki/open-city-profile-ui/pull/154#discussion_r618158453

VilleMiekkoja commented on src/common/test/myProfileMocking.ts:
```diff
@@ -155,3 +207,189 @@ export const getVerifiedData = (
+    getProfile() {
+      return currentProfile;
+    },
+    add(dataType, data, index = -1) {
```
This seems to be missing typings?

NikoHelle replied:
Not needed, because TS can combine returned object and function's return type (ManipulationFunctions)

VilleMiekkoja commented on src/common/test/myProfileMocking.ts:
```diff
@@ -155,3 +207,189 @@ export const getVerifiedData = (
+    getFormValues() {
+      const basicData = pickSources(
+        currentProfile,
+        basicDataType
```
Does basicDataType refer to first name, last name and nickname? Would _personNamesDataType_ describe it better?

NikoHelle replied:
They are grouped in UI as "Perustiedot". That's why using "Basic data". It might get additional fields that are not names. Language was in it too, but moved to "Lisätiedot".

VilleMiekkoja commented on src/common/test/myProfileMocking.ts:
```diff
@@ -155,3 +207,189 @@ export const getVerifiedData = (
+  Object.assign(profile, { [propName]: newPrimary });
+}
+
+export function cloneAndManipulateProfile(
```
I find the naming of this function little problematic. It does clone the profile, but does not really give the result without calling the additional functions that it returns. It feels more like a class type of object than a function that does a particular thing. Maybe renaming it to _clonedProfileWithManipulationFunctions_ etc would be more describing?

NikoHelle replied:
"clonedProfileWithManipulationFunctions" sounds like a return type?

Renamed to cloneProfileAndProvideManipulationFunctions (which is a long monster :D ) IMHO the actual return type describes the function already quite well. But that is only seen via hover in the IDE.

VilleMiekkoja commented on src/graphql/generatedTypes.ts:
```diff
@@ -608,9 +605,42 @@ export interface UpdateMyProfile_updateMyProfile_profile_primaryAddress {
+  readonly addressType: AddressType | null;
+}
+
+export interface UpdateMyProfile_updateMyProfile_profile_addresses_edges_node {
```
The amount of different naming conventions this interface applies is ambitious to say the least 😅 . These graphql types seems to be automatically generated? Could we configure the naming conventions somehow? There also seems to be some repetition with the word _UpdateMyProfile_ that I'm not sure if it provides any additional value?

NikoHelle replied:
Yes, there is codegen that fetches these monsters from graphQL server and UI cannot rename these. Typings.ts was added to make common "shortcuts".

VilleMiekkoja commented on src/profile/components/basicData/BasicData.tsx:
```diff
@@ -1,28 +1,172 @@
+  } = useNotificationContent();
   const testId = basicDataType;
 
   if (!editDataList || !editDataList[0]) {
```
This seems to be the first row where return values from _useProfileDataEditor_ function call are being applied. I wonder if it would increase readability if we make the function call _useProfileDataEditor_ straight above this line 45, so that it would be closer to the place where it is used?

NikoHelle replied:
No, because React hooks ("useNotificationContent") cannot be run conditionally.

VilleMiekkoja replied:
I meant that useProfileDataEditor function call would be above this condition, at line 44

NikoHelle replied:
Ah... Done.

VilleMiekkoja commented on src/profile/components/basicData/BasicData.tsx:
```diff
@@ -1,28 +1,172 @@
+    fieldName: keyof FormikValues
+  ): boolean => getIsInvalid<FormikValues>(formikProps, fieldName, true);
+
+  const getFieldErrorMessage = (
```
I find the naming of this function little problematic. It is that it does not seem to return the FieldErrorMessage if there is no fieldError? One option would be to either move the check for _hasFieldError_ outside of this function, and then only call this function when we are sure it will return the FieldErrorMessage. Other option would be to name it differently: _getFieldErrorMessageIfExists_ or _getFieldErrorMessageOrUndefined_. Those names would give hint to user of the function that there is a chance it won't return the FieldErrorMessage

NikoHelle replied:
Isn't typescript return types enough to tell what function returns? :) In this case it is not written here but seen when you hover over the function. And where ever the function is called, you can see return types by hovering over the function call. This is editor specific of course, but arent' they visible in all normal editors?

I find it hard to put all return types to the function name itself.

VilleMiekkoja replied:
Ide and hovering definitely helps. But I find it easier to read if I can determine what the function does (returns in this case) from the name. Hovering over is an extra step that slows down reading, and should be only needed if I wish to know more details. I agree that a function with many return types in its name feels bad. I think in general it would be good for a function to only return one types of things. The more types the function returns, the more complex it is, and the more harder it is for the reader of the code. I would probably prefer moving the check for hasFieldError outside of this function for that reason.  But yeah obviously this is very opinion dependant when it comes to readability and what is enough.

NikoHelle replied:
Two reasons why hasFieldError is in getFieldErrorMessage

1: it is used in an element-prop:
helperText={getFieldErrorMessage(formikProps, 'nickname')}

in this case it should be changed to:
helperText={hasFieldError(formikProps, 'nickname') && getFieldErrorMessage(formikProps, 'nickname')}

that's looks terrible IMHO. And that in 3 places in this file and later on in address, postal code, city, email and phone. 

2: it has always been there. It this file it looks like it was created here, but it's a copy from old code

NikoHelle replied:
Should I copy the return type of getFieldError to getFieldErrorMessage? Copying code for faster readability is not very convenient. When it getFieldError changes, it must be changed everywhere. Later on also in addresses, emails and phones will use it.

I can make a common type for them, but seems a bit excessive. You cannot read it in IDE either.

VilleMiekkoja replied:
To my eye this:

```helperText={hasFieldError(formikProps, 'nickname') && getFieldErrorMessage(formikProps, 'nickname')}```

doesn't look that bad because you then know the condition when there will be a helperText. 😅 But I guess this is about what reader of the code is accustomed to. I've probably just used to read code where logic is shown in higher abstraction level. I think the current solution is good enough, we can go with that. 👍

NikoHelle replied:
I left this as it is.

VilleMiekkoja commented on src/profile/components/basicData/BasicData.tsx:
```diff
@@ -1,28 +1,172 @@
+  const getFieldErrorMessage = (
+    formikProps: FormikProps<FormikValues>,
+    fieldName: keyof FormikValues
+  ) => {
```
This seems to be missing typings for the return value?

NikoHelle replied:
TS compiler / linter and Sonar would alert if needed. TS understands the return type is the return type of the  function called: "return getFieldError(....)"

VilleMiekkoja commented on src/profile/components/basicData/BasicData.tsx:
```diff
@@ -1,28 +1,172 @@
+  };
+
+  const onAction: ActionListener = async (action, item, newValue) => {
+    clearMessage();
```
Would it be good idea to move the _useNotificationContent_ function call to line 67, so clearMessage used here would be defined closer to usage?

NikoHelle replied:
Referring to another re-ordering comment:
https://github.com/City-of-Helsinki/open-city-profile-ui/pull/154#discussion_r618158453

VilleMiekkoja commented on src/profile/components/basicData/BasicData.tsx:
```diff
@@ -1,28 +1,172 @@
+    setSuccessMessage,
+    clearMessage,
+  } = useNotificationContent();
   const testId = basicDataType;
```
Im not sure if renaming the basicDataType to testId provides additional value? Could we apply the basicDataType directly instead?

NikoHelle replied:
Done

VilleMiekkoja commented on src/profile/components/basicData/__tests__/BasicData.test.tsx:
```diff
@@ -26,36 +35,157 @@ describe('<BasicData /> ', () => {
   };
+  const t = i18n.getFixedT('fi');
+  const editButtonSelector: ElementSelector = { id: 'basic-data-edit-button' };
+  const cancelButtonSelector: ElementSelector = {
```
cancelButtonSelector seems to be only used in the last test. Would it increase readability to move it inside the last test, as it seems not to be needed elsewhere yet?

NikoHelle replied:
I thought it was cleaner to group selectors in one place. I'll change this.

NikoHelle replied:
Done.

VilleMiekkoja commented on src/profile/components/basicData/__tests__/BasicData.test.tsx:
```diff
@@ -1,22 +1,31 @@
-  const renderTestSuite = (responses: MockedResponse[]) => {
+  const responses: MockedResponse[] = [];
+  const initialProfile = getMyProfile().myProfile as ProfileData;
+  const renderTestSuite = () => {
```
I would probably move this closer to usage, to line 92, above _initTests_ function. Same with initialProfile  object above at line 27.

NikoHelle replied:
Referring to another re-ordering comment:
https://github.com/City-of-Helsinki/open-city-profile-ui/pull/154#discussion_r618158453

VilleMiekkoja commented on src/profile/components/basicData/__tests__/BasicData.test.tsx:
```diff
@@ -26,36 +35,157 @@ describe('<BasicData /> ', () => {
+  });
+  it('sends new data and returns to view mode when saved', async () => {
+    // new values
+    const basicData = {
```
This _basicData_ initialisation seems to be made in the beginning of each test. I wonder if we could initialise it inside the _beforeEach_?

NikoHelle replied:
Done.

VilleMiekkoja commented on src/profile/components/editingNotifications/EditingNotifications.module.css:
```diff
@@ -0,0 +1,10 @@
+  position: relative;
+}
+.wrapper:focus {
+ outline:none;
```
Indentation and spacing seems little off. Think it would be a good idea to setup automated css lint script that would catch these. Maybe also prettier that would automatically format these correctly. Hauki applies stylelint library for this: https://www.npmjs.com/package/stylelint

NikoHelle replied:
Ah, we have prettier. My IDE shows an error with the plugin, so it was not run.

NikoHelle replied:
Ran all css (old and new) through prettier. All changed!

VilleMiekkoja commented on src/profile/components/editingNotifications/EditingNotifications.tsx:
```diff
@@ -0,0 +1,35 @@
+  content,
+  dataType,
+}: Props): React.ReactElement | null {
+  if (!content.text) {
```
According to clean code passing or returning null is questionable: https://enos.itcollege.ee/~jpoial/oop/naited/Clean%20Code.pdf#page=141

Because then you have to do these checks in the beginning of components and they don't render what you would expect them to render. Maybe one option is to check the content.text value in BasicData.tsx, and only render this EditingNotifications if it has a value 🤔

NikoHelle replied:
I think it is normal in React world to render nothing (=return null), if component don't have correct data etc. to render its content. Developer can see it in return type definition.

This is a common component which will be used by also by address, phones, emails editors. It is easier and clearer to check content here than every time component is used. 

BasicData:
{ if !context.text && <EditingNotifications .../> }

Addresses:
{ if !context.text && <EditingNotifications .../> }

and so on...

And if there is a new prop (content.alert) that must be added to each implementation instead of in the component itself.

{ if !context.text && !content.alert && <EditingNotifications .../> }

VilleMiekkoja commented on src/profile/components/editingNotifications/useNotificationContent.ts:
```diff
@@ -0,0 +1,65 @@
+      : t('notification.genericSuccess');
+  };
+  const setSuccessMessage: ContentHandlers['setSuccessMessage'] = (
+    message,
```
Is this message used? In BasicData it feels like it is always set to an empty string. For example in line 41, it seems the callers always use action, so the result is never the message directly? I wonder if it would simplify the code if we removed that unused parameter?

NikoHelle replied:
Good point, it was added for one situation which does not exist anymore. Will be removed

NikoHelle replied:
Done.

VilleMiekkoja commented on src/profile/components/formButtons/FormButtons.tsx:
```diff
@@ -0,0 +1,50 @@
+    <div
+      className={classNames([
+        commonFormStyles.editButtons,
+        commonFormStyles.editButtons,
```
Hmm.. is this twice on purpose?

NikoHelle replied:
No :) Will be fixed.

NikoHelle replied:
Done.

VilleMiekkoja commented on src/profile/components/saveIndicator/SaveIndicator.tsx:
```diff
@@ -0,0 +1,31 @@
+const SaveIndicator = (props: Props): React.ReactElement | null => {
+  const { action, testId } = props;
+  const { t } = useTranslation();
+  if (!action) {
```
I would prefer this check to be in BasicData.tsx, but it can be just a preference thing. If you render the SaveIndicator component now, you can't be sure that it actually renders what it says, but it can return null also. I just think that the logic for whether we should show SaveIndicator or not belongs to a higher level than to the SaveIndicator component itself. Or then maybe one option could be to make the api of the SaveIndicator more clear, something like _showSaveIndicator_ prop that would indicate if it's rendered more easily. Now users of this component needs to dig in to the implementation of the component to see if it's actually rendered.

NikoHelle replied:
Same comment than above: components that are used by multiple components, should make these check internally in once place instead of in all components that use it and it's ok to return null.

https://github.com/City-of-Helsinki/open-city-profile-ui/pull/154#discussion_r617231848

VilleMiekkoja commented on src/profile/components/saveIndicator/SaveIndicator.tsx:
```diff
@@ -0,0 +1,31 @@
+  return (
+    <div
+      role="alert"
+      className={commonFormStyles.visuallyHidden}
```
Is the SaveIndicator hidden from screen?

NikoHelle replied:
When form is submitted, buttons are disabled while saving. SaveIndicator is just for screen readers to indicate that submit happened and what kind of action (save/remove) is being done.

VilleMiekkoja commented on src/profile/helpers/__tests__/editData.test.ts:
```diff
@@ -102,5 +165,187 @@ describe('editData.ts ', () => {
+        expect(newTarget.value).toEqual(newValue);
+        // ...and in formValues
+        expect(newFormValue).toMatchObject(newValue);
+        // new old old id and primary match
```
new id and primary match with old one?

NikoHelle replied:
Typo fixed.

VilleMiekkoja commented on src/profile/helpers/__tests__/editData.test.ts:
```diff
@@ -102,5 +165,187 @@ describe('editData.ts ', () => {
         });
       }
     });
+    it(`updateEditData clones and updates data when dataType is ${dataType} and returns formValues`, () => {
```
Is the function being tested updateItemAndCreateSaveData? I wonder if updateEditData in this sentence should be replaced with updateItemAndCreateSaveData?

NikoHelle replied:
Fixed typo

VilleMiekkoja commented on src/profile/helpers/__tests__/editData.test.ts:
```diff
@@ -102,5 +165,187 @@ describe('editData.ts ', () => {
+      expect(updatedItem).toEqual({ ...item, saving: undefined });
+      expect(updatedItem.saving).toEqual(undefined);
+    });
+    it(`Added new items are not lost and are updated from source'`, () => {
```
Should we skip this test for now or add todo or jira id? It might be confusing looking test results saying this test pass when it does not do anything yet?

NikoHelle replied:
Ok, I left it there as a placeholder. Adding new items is in next PR. I'll delete this.

NikoHelle replied:
Removed

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -80,17 +87,42 @@ export type EditData = {
 };
 
-function isMultiItemDataType(dataType: EditDataType): boolean {
+export function isMultiItemDataType(dataType: EditDataType): boolean {
```
Is this export necessary? Tried searching but did not find any usages outside of this file?

NikoHelle replied:
Removed

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -80,17 +87,42 @@ export type EditData = {
+
+export function hasNewItem(allItems: EditData[]): boolean {
+  return !!getNewItem(allItems);
+}
```
Did not find any usages of these functions outside of this file, so the exports are probably not needed?

NikoHelle replied:
Correct, they (like other unused exports) were exported for testing, but they are tested via functions that use them.

NikoHelle replied:
Removed

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -127,7 +159,7 @@ function create(
   };
```
Comment for line 155: According to my ide the line could be simplified to:

`primary: (profileData as MultiItemProfileNode).primary`

NikoHelle replied:
Changed. Affected tests and fixed them.

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -105,15 +137,15 @@ function getValueProps(dataType: EditDataType): string[] {
 }
 
-function create(
+function createNewItem(
```
The function definition (line 148) is pretty far away from where it is used first time in this file: line 303. I wonder if it would be possible to move it closer to usage and if it would increase readability?

NikoHelle replied:
Referring to another re-ordering comment:
https://github.com/City-of-Helsinki/open-city-profile-ui/pull/154#discussion_r618158453

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -127,7 +159,7 @@ function create(
 }
 
-function pickSources(
+export function pickSources(
```
First usage of this function inside this file is pretty far away from the definition 162 -> 362. I wonder if it's possible to move it closer to usage to help readability? Sometimes can be hard if there's many functions that if you move one function closer, then some other function will be more far away 😅

NikoHelle replied:
Referring to the Team's Slack discussion, ordering functions and keeping them in logical order for everyone is a tedious job and hard to keep the ordering in the long run.

We should write the code so functions are in clear, logical order and try to keep them so, but it is not required and enforced.

VilleMiekkoja commented on src/common/test/myProfileMocking.ts:
```diff
@@ -155,3 +207,189 @@ export const getVerifiedData = (
   ...overrides,
 });
+
+function createNewProfileNode(dataType: EditDataType): MultiItemProfileNode {
```
There seems to be almost similar function defined in editData.ts line 227:

`export function createNewProfileNode<T extends MultiItemProfileNode>(...`

I wonder if we could apply it here some way?

NikoHelle replied:
Good catch!

NikoHelle replied:
Done.

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -150,17 +182,254 @@ function pickSources(
   }
 }
 
+export function cloneAndMutateItem(
```
Did not find any usages outside of this file, so maybe the export is unnecessary?

NikoHelle replied:
Removed.

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -150,17 +182,254 @@ function pickSources(
+  };
+}
+
+export function findItemIndex(
```
Did not find any usages outside of this file, so maybe the export is unnecessary?

NikoHelle replied:
Removed.

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -150,17 +182,254 @@ function pickSources(
+  return allItems.find(item => id === item.id);
+}
+
+export function updateItemAndCloneList(
```
Did not find any usages outside of this file, so maybe the export is unnecessary?

NikoHelle replied:
Removed.

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -150,17 +182,254 @@ function pickSources(
+  return newList;
+}
+
+export function createNewProfileNode<T extends MultiItemProfileNode>(
```
Did not find any usages outside of this file, so maybe the export is unnecessary? Though there's one similar looking function at myProfileMocking.ts line 211, maybe these could be combined?

NikoHelle replied:
Yes, I merged them.

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -150,17 +182,254 @@ function pickSources(
+  };
+}
+
+export function createFormValues(
```
Did not find any usages outside of this file, so maybe the export is unnecessary?

NikoHelle replied:
Removed.

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -150,17 +182,254 @@ function pickSources(
+  return dataHasUpdated ? newList : null;
+}
+
+export function createBackups(): Backups {
```
Did not find any usages outside of this file, so maybe the export is unnecessary?

NikoHelle replied:
Removed.

VilleMiekkoja commented on src/profile/helpers/updateMutationVariables.ts:
```diff
@@ -194,33 +199,48 @@ function formMutationArrays<T extends AddressNode | EmailNode | PhoneNode>(
+  };
+};
 
 export { formMutationArrays, updateMutationVariables };
```
It seems in some files the exports are inlined, and in some files like this they are at the end of the file. I wonder if it would be good idea to use one style of exporting mainly? Inlined exports seems to be more common maybe?

VilleMiekkoja replied:
Also did not find usages for _formMutationArrays_. Maybe exporting it is unnecessary?

NikoHelle replied:
This is an old file made by someone else. Haven't followed their preference of exporting. I think exporting where the function actually is, is much better. Won't find all occurences and change them to match that 😄

VilleMiekkoja commented on src/profile/helpers/updateMutationVariables.ts:
```diff
@@ -1,4 +1,4 @@
-import { isEqual } from 'lodash';
+import _, { isEqual } from 'lodash';
```
Should we import the pick separately like in the case of isEqual? Or just '_' to handle both?

NikoHelle replied:
Auto-import did this. _.isEqual is better.

NikoHelle replied:
Done.

VilleMiekkoja commented on src/profile/hooks/useProfileDataEditor.ts:
```diff
@@ -1,34 +1,109 @@
+  reset: (item: EditData) => void;
 };
 
+export type Action = SaveType | 'edit' | 'cancel' | 'save' | 'add';
```
What's the difference between SaveType and 'save'? Sounds semantically similar.

NikoHelle replied:
SaveType in EditData tells what property is being saved: 'values' or 'set-primary'.  'save' is just an general action.

VilleMiekkoja commented on src/profile/hooks/useProfileDataEditor.ts:
```diff
@@ -1,34 +1,109 @@
+    resetItem,
+  } = useMemo(
     () => createEditorForDataType(profileData, dataType),
     // eslint-disable-next-line react-hooks/exhaustive-deps
```
Why do we not want to recompute if profileData or dataType has changed?

NikoHelle replied:
Don't want to override current data. It may include new data items (in case of addresses, emails, phones) or unsaved data that are not in profileData. Update is done separately in useEffect.

VilleMiekkoja commented on src/profile/hooks/useProfileDataEditor.ts:
```diff
@@ -1,34 +1,109 @@
 
+  const [, setUpdateTime] = useState<number>(Date.now());
+  const triggerUpdate = () => {
+    setUpdateTime(Date.now());
```
Not sure what is the purpose of this triggerUpdate? Is it to just manually trigger UI update by applying useState? It kind of feels like a hacky thing to do. I wonder if there's any alternatives 🤔

NikoHelle replied:
Yes, when data is updated in some way that does not make hook re-render (like pressing cancel) the re-render must be triggered manually. Same kind of easy re-render is used for example in MobX:

https://github.com/mobxjs/mobx-react-lite/blob/master/src/utils/utils.ts

VilleMiekkoja commented on src/profile/hooks/useProfileMutations.ts:
```diff
@@ -0,0 +1,145 @@
+import { updateMutationVariables } from '../helpers/updateMutationVariables';
+import { ProfileContext, ProfileContextData } from '../context/ProfileContext';
+
+const UPDATE_PROFILE = loader('../../profile/graphql/UpdateMyProfile.graphql');
```
Is there a reason why these follow different naming conventions than regular const variables?

NikoHelle replied:
Only reason is that the old Profile UI had them like that the code is moved here.

VilleMiekkoja commented on src/profile/hooks/useProfileMutations.ts:
```diff
@@ -0,0 +1,145 @@
+  '__typename' | 'id'
+>;
+
+export type ValuesOfUpdateProfileData = UpdateProfileData[UpdatableProfileDataKeys];
```
Did not find any usages of this type outside of this file, so maybe the export is unnecessary?

NikoHelle replied:
Removed.

VilleMiekkoja commented on src/profile/hooks/useProfileMutations.ts:
```diff
@@ -0,0 +1,145 @@
+    return updateProfile({ variables });
+  };
+
+  return { update, error, loading, profileData };
```
Did not find usages for error and loading, maybe there's no need to return those?

NikoHelle replied:
Except in useProfileMutations.test.ts. UI does not use them because Promise.reject() is catched instead. But in tests error and loading are directly tested. Should they be kept there?

VilleMiekkoja replied:
Are there plans on using them in the UI in future? If not, it might simplify things to delete them if they are not needed.

NikoHelle replied:
Ok, I removed them and changed the test just to  check error wont alter profile data

VilleMiekkoja commented on src/profile/hooks/useProfileMutations.ts:
```diff
@@ -0,0 +1,145 @@
+    profileData &&
+    cache.identify((profileData.myProfile as unknown) as Reference);
+  if (!updatedProfile || !identity) {
+    throw new Error('Unable to update cache2');
```
cache2? This might be that Im beginner related to graphql, but is there cache1 somewhere?

NikoHelle replied:
That's a relic from some debugging session. :)

VilleMiekkoja commented on src/profile/helpers/__tests__/updateMutationVariables.test.ts:
```diff
@@ -106,112 +108,251 @@ const newAddress = {
-  expect(variables.input.profile.updatePhones).toEqual([
-    getPhoneAsComparisonObject(updatedSecondaryPhone),
-  ]);
+  test('Variables in update data match formValues even if not changed', () => {
```
There seems to be 2 different conventions for this test function name: for example editData.test.ts applies function _it_. I wonder if it would be a good idea to only use one of them?

NikoHelle replied:
Yes, "test"-funcs are used in the old project files. I kept the same func calls inside each file. I'll rename them.

NikoHelle replied:
Done.

VilleMiekkoja commented on src/profile/helpers/__tests__/updateMutationVariables.test.ts:
```diff
@@ -106,112 +108,251 @@ const newAddress = {
-    getPhoneAsComparisonObject(updatedSecondaryPhone),
-  ]);
+  test('Variables in update data match formValues even if not changed', () => {
+    const variables: UpdateMyProfileVariables = updateMutationVariables(
+      formValues,
+      myProfile
+    );
```
This variables initialisation seems to be done in multiple tests (in 7 total). Not sure if it could be initialised in beforeEach, and then in tests that do not wish to use it, they would reinitialise it differently. Though sometimes repetition actually improves readability it feels, so not sure 🤔

NikoHelle replied:
Yeah, there are 7 of those code blocks and in between 5 that does different call. That's why the call is not in beforeEach. Those 7 cannot be logically grouped under one describe(). I thought about using a helper function, but then the readability came to mind. Would it be harder to follow the test if some tests had different initialisation function than others?

Instead of 

```
const variables: UpdateMyProfileVariables = updateMutationVariables(
      formValues,
      myProfile
    );
```
it was something like

`createCommonUpdateVariables();`

Or create one helper function for all tests - with optional formValues:

`createUpdateVariables(formValues?:FormValues);`

Better or worse? :)

VilleMiekkoja replied:
One option could be to make 1 more master describe on top of those all, and inside that there would be beforeEach. Not sure if it would make things more complicated though. Repetition of those initialisations like it is now might be best solution for readability

NikoHelle replied:
Left as it is.

VilleMiekkoja commented on src/profile/helpers/__tests__/updateMutationVariables.test.ts:
```diff
@@ -106,112 +108,251 @@ const newAddress = {
-  expect(variables.input.profile.updateAddresses).toEqual([]);
-  expect(variables.input.profile.updateEmails).toEqual([]);
-  expect(variables.input.profile.updatePhones).toEqual([]);
+describe('Update variable for language is formed correctly', () => {
```
Minor notice about language: It seems in some of the tests the first letter is capitalised and in some tests it's not. Not a big thing though

NikoHelle replied:
Yes, sometimes it's a typo, usually the case is:

describe("This describes the purpose of the test group ')
it('and this is in lower case becasue it continues the sentence started in describe()")

VilleMiekkoja commented on src/profile/helpers/__tests__/updateMutationVariables.test.ts:
```diff
@@ -106,112 +108,251 @@ const newAddress = {
-  expect(variables.input.profile.removeAddresses).toBeFalsy();
-  expect(variables.input.profile.removeEmails).toBeFalsy();
-  expect(variables.input.profile.removePhones).toBeFalsy();
+  test('but empty update is also handled', () => {
```
This feels like a side sentence to another sentence? It might be a good idea to make each test and its name independent of other test names for clarity. If the other test is then removed or modified, this one might not make as much sense anymore. Maybe just remove the first word _but_?

NikoHelle replied:
Fixed.

VilleMiekkoja commented:
Tested it locally and seems to be working nice! Coding this has required lot of effort, well done! 🎖️

NikoHelle committed:
Unify test files - always use "it()"
Some tests used test(), some it(). Now all use it()

NikoHelle committed:
Ran prettier for all .css files

For some prettier extension issue, css files had never been run through prettier.

NikoHelle committed:
Identical functions merged
myProfileMocking had identical function with EditData.
Removed from myProfileMocking.

NikoHelle committed:
Removed useless testId from BasicData

NikoHelle committed:
Decrease test data repetition

Common code grouped as one and removed useless const

SQ basicdata.test. update

NikoHelle committed:
Renamed cloneAndManipulateProfile to
cloneProfileAndProvideManipulationFunctions

NikoHelle committed:
Renamed "triggerAction" to "clickElement" in tests

NikoHelle committed:
Renamed test function to describe it better

renderProfileContextWrapper renamed
 to renderComponentWithMocksAndContexts

NikoHelle commented:
All changes should be done now. Didn't make function re-orderings.

VilleMiekkoja reviewed (approved):
Well done! 👍
